std/compress/internal/io.go

```diff
@@ -0,0 +1,226 @@
+	nr.next(v)
+}
+
+func (nr *NumReader) next(v frontend.Variable) frontend.Variable {
+	if len(nr.c) == 0 {
+		return 0
+	}
+
+	if nr.last == nil { // the very first call
+		nr.last = compress.ReadNum(nr.api, nr.c, nr.wordsPerNum, nr.stepCoeff)
+		if v != nil {
+			nr.api.AssertIsEqual(nr.last, v)
+		}
+		return nr.last
+	}
+
+	nr.last = nr.api.Sub(nr.api.Mul(nr.last, nr.stepCoeff), nr.api.Mul(nr.c[0], nr.maxCoeff))
+	if nr.wordsPerNum < len(nr.c) {
+		if v == nil {
+			nr.last = nr.api.Add(nr.last, nr.c[nr.wordsPerNum])
+		} else {
+			addPlonkConstraint(nr.api, nr.last, nr.c[nr.wordsPerNum], v, 1, 1, -1, 0, 0)
+			nr.last = v
+		}
+	} else if v != nil {
+		panic("todo refactoring required")
+	}
+
+	nr.c = nr.c[1:]
+	return nr.last
+}
+
+// TODO Use std/rangecheck instead
+type RangeChecker struct {
+	api    frontend.API
+	tables map[uint]*logderivlookup.Table
+}
+
+func NewRangeChecker(api frontend.API) *RangeChecker {
+	return &RangeChecker{api: api, tables: make(map[uint]*logderivlookup.Table)}
+}
+
+func (r *RangeChecker) AssertLessThan(bound uint, c ...frontend.Variable) {
+
+	var check func(frontend.Variable)
+	switch bound {
+	case 1:
+		check = func(v frontend.Variable) { r.api.AssertIsEqual(v, 0) }
+	case 2:
+		check = r.api.AssertIsBoolean
+	case 4:
+		check = r.api.AssertIsCrumb
+	default:
+		cRangeTable, ok := r.tables[bound]
+		if !ok {
+			cRangeTable := logderivlookup.New(r.api)
+			for i := uint(0); i < bound; i++ {
+				cRangeTable.Insert(0)
+			}
+		}
+		_ = cRangeTable.Lookup(c...)
+		return
+	}
+	for i := range c {
+		check(c[i])
+	}
+}
+
+// LessThan returns a variable that is 1 if 0 \leq c < bound, 0 otherwise
+// TODO perf @Tabaie see if we can get away with a weaker contract, where the return value is 0 iff 0 \leq c < bound
+func (r *RangeChecker) LessThan(bound uint, c frontend.Variable) frontend.Variable {
+	switch bound {
+	case 1:
+		return r.api.IsZero(c)
+	}
+
+	if bound%2 != 0 {
+		panic("odd bounds not yet supported")
+	}
+	v := plonk_helpers.EvaluatePlonkExpression(r.api, c, c, -int(bound-1), 0, 1, 0) // c^2 - (bound-1)*c
+	res := v
+	for i := uint(1); i < bound/2; i++ {
+		res = plonk_helpers.EvaluatePlonkExpression(r.api, res, v, int(i*(bound-i-1)), 0, 1, 0)
+	}
+
+	return r.api.IsZero(res)
+}
+
+var wordNbBitsToHint = map[int]hint.Hint{1: BreakUpBytesIntoBitsHint, 2: BreakUpBytesIntoCrumbsHint, 4: BreakUpBytesIntoHalfHint}
+
+// BreakUpBytesIntoWords breaks up bytes into words of size wordNbBits
+// It also returns a Slice of bytes which are a reading of the input byte Slice starting from each of the words, thus a super-Slice of the input
+// It has the side effect of checking that the input does in fact consist of bytes
+func (r *RangeChecker) BreakUpBytesIntoWords(wordNbBits int, bytes ...frontend.Variable) (words, recombined []frontend.Variable) {
+
+	wordsPerByte := 8 / wordNbBits
+	if wordsPerByte*wordNbBits != 8 {
+		panic("wordNbBits must be a divisor of 8")
+	}
+
```

Comment:
I would be interested in comments here, a general description of the method; 
as far as I understand, the "naive" way to do that is simply to do bit decomposition of the input bytes and recompose them with granularity we are lookin for (`wordNbBits`). 
How much better is your method vs this and why?